Adding  image gallery

rdquintas opened the pull request:
Hi guys
Image Gallery is concluded

This relates to this issue
https://github.com/lnug/lnug.github.io/issues/97

Please let me know what you think.

Here's the summary of what I did
- images are stored inside folder images/gallery
- every time you run npm build, markup is created for page image-gallery (this is a new page btw)
- inserted "Gallery" link on the footer
- I have also added 3 images to the home page. these 3 images change randomly. I think this adds some dynamic to the website (this was implemented using Javascript)
- I have added 2 JS libraries: one to handle the image viewing when you click, the other to handle the lazy-loading (this is important for mobile devices)

To-Dos and future Improvements
- There's no thumbs. maybe we could use a npm module to create those with npm build (I have already done some work around this)
- We could fetch images automatically from Twitter, Instagram, etc... using node. Not sure if this would be a good idea.
- add metadata to each image. I was thinking of adding a little JSON file to each image so that we could have important information such as: author, origin, date, etc...

Thanks

rdquintas committed:
adding image gallery

rdquintas committed:
inserting new page for image gallery

rdquintas committed:
adding image-gallery main logic

rdquintas committed:
json logic to handle image loading

rdquintas committed:
adding js libs jquery and lazy loading

rdquintas committed:
fixing css

rdquintas committed:
added Magnific Popup js

rdquintas committed:
adding images and improve layout with Masonry

rdquintas committed:
removing masonry and adding home page html

rdquintas committed:
final touches

rdquintas committed:
finishing README etc...

simonmcmanus commented:
Hi @rdquintas ,

Thanks for this PR!

Currently this PR is not linting correctly. 

You can run this command locally to test the linting:

npm test

We use the Standard js library for linting, you can read more here: 

https://github.com/feross/standard

I will try to pull this down later today to have a play but we deffo need to fix the linting before it can be merged. 

If you have any questions please let me know.

rdquintas commented:
Thanks @simonmcmanus ,will do that and submit new pull request

simonmcmanus commented:
No need to create a new PR. If you just fix on your current branch and push it will update this PR.

simonmcmanus commented on image-gallery.html:
```diff
@@ -0,0 +1,284 @@
+    <h2 class="url"></h2>
+    <h2 class="description"></h2>
+    <h2 class="author"></h2>
+    <h2 class="date"></h2>
```
what are all these empty tags doing here?

You should either populate them in lib/image-gallery.js or drop them from the template

rdquintas replied:
@simonmcmanus you're right
forgot to clean that up, sorry

will fix it

simonmcmanus commented:
Generally this is great, you have looked at the existing app and seen how it fits together. 

I'm not convinced we need magnific-popup. Its adding to the page weight with css and and js but not adding much value from what I can tell. 

If we add a library for every feature the website will get bloated very quickly.  I'd just open the full image in a new tab.

When I turn JS off its not showing the images on the gallery page, I think thats just magnific-popup causing issues.

I also think we need to add thumbnails before we can merge this in. Loading three full size images on the homepage is just too much, this gallery page is also going to be massive. 

I'd suggest at build time we create square thumbnails that we import into repo (we dont need the full photos in this repo) 

I think the thumnails need to be square to avoid issues like this: 

![screen shot 2016-02-12 at 10 21 33](https://cloud.githubusercontent.com/assets/55853/13004280/81e5b68e-d172-11e5-9c38-fed72446f52d.png)
![screen shot 2016-02-12 at 10 21 39](https://cloud.githubusercontent.com/assets/55853/13004281/83c77c30-d172-11e5-831b-eed546603a63.png)
![screen shot 2016-02-12 at 10 21 33](https://cloud.githubusercontent.com/assets/55853/13004284/867cd1f0-d172-11e5-8cff-86a4f7d5491f.png)

I've not tried it yet but project Oxford might help us to generate some nice thumbnails:

https://www.projectoxford.ai/vision

Need to see what we can do on the free teir tho.

Also, is data/image-gallery.json hard coded at the moment? I couldn't see any script to update it. 

Any questions just give me a shout.

rdquintas committed:
removing empty tags

rdquintas committed:
removing magnific-popup  :o(

rdquintas commented:
@simonmcmanus Thanks for the awesome feedback!
- regarding thumbnails, I've been fiddling with this https://www.npmjs.com/package/node-thumbnail and works fine. can we use it ? (requires imagemagic to be installed on your machine thought)
- one thing I didn't fully understood yet: do you want to have the photos inside the repo or not  ?@iancrowther mentioned to me that the photos would be stored on the repo. So I'm confused.
- I agree with your comments on magnific popup lib. I'll remove it. no prob.
- I also thought the page will be massive if we start adding tons of pics, hence I added a lazy-load JS lib to manage that.
  even if we only have thumbs, if you still have a single page with, say, 1000 thumbs, that will be a problem.
  so I thought a lazy-load mechanism would help.
  but if you want this page to work with javascript disabled, how should we tackle the scrolling/viewing of a massive gallery web page ? maybe a different pagination mechanism ? what do you suggest ?

Cheers.

iancrowther commented:
@simonmcmanus @rdquintas I said for the first draft, commit the images to keep things lean.

We can automate later

iancrowther commented:
@rdquintas I agree on page size.  Let's get the UI/UX right, fix the nav discussions etc

iancrowther commented:
looks like the linting needs some tlc

rdquintas committed:
fixing linting

rdquintas committed:
fix image element. opens in separate TAB

rdquintas committed:
removing jquery and lazy-load.js

rdquintas committed:
repplaced all my jQuery code with vanilla Javascript

rdquintas committed:
forgot to build markup

rdquintas committed:
updating README

rdquintas committed:
linting build-markup,js

rdquintas committed:
linting app.js

rdquintas commented:
Nice.  
It seems all checks have passed now.
:o)

Here's a summary

## This is Done
- removed empty tags
- linted all my code using standard.js
- images now open on separate tab
- removed all external libs (magnify-popup, lazy-load and jQuery)
- I'm  now using vanilla javascript instead

## This it NOT
- use https://www.npmjs.com/package/node-thumbnail to create thumnails
- adapt UI/UX to only display "square" photos

rdquintas committed:
removing file magnific-popup.css

simonmcmanus commented:
Thanks @rdquintas 

I'd like to see the thumbnails in this repo but not the main pics. If we only show the thumbnails on the site we don't need to full pics.

re paging, as it's currently hardcoded let's get it looking right with the pictures we have in place at the moment and then worry about loading more pics later.  We have a number of interesting options. :)

node-thumbnail look fine. That can be a separate PR though, for the time being you can just replace the big pics with the thumbnails.

Also, on the homepage I think clicking on one of the pictures should take you to the full size picture not the gallery. We could have a button underneath the pics say [See gallery]

rdquintas commented:
Hi @simonmcmanus thanks again.

Sure no problem, I will apply those changes.

Regarding the original images then, do you think it would be a good idea to have a JSON file like this for example

```
[{
    url: "http://www.twitter.com/werwerwefojhwof.jpg",
    name: "image1",
    author: "AAAAAA",
    date: "20150402"
},{
    url: "http://www.twitter.com/werwerwefojhwof.jpg",
    name: "image1",
    author: "AAAAAA",
    date: "20150402"
},{
    url: "http://www.twitter.com/werwerwefojhwof.jpg",
    name: "image1",
    author: "AAAAAA",
    date: "20150402"
}]
```

and we would then use it to build our markup with `npm build`

The file itself would be changed manually (for now at least).

Whad to you trhink :?

Cheers

rdquintas committed:
adding node-easyimage to process thumbs

rdquintas committed:
final CSS touches. adding button VIEW GALLERY

rdquintas commented:
Hi @simonmcmanus 
Here's another update
- added new **thumbs** folder
- all images have now their corresponding thumbnail on that same folder
- added a new build instruction called `make-thumbs` it will generate thumbs for all the images stored inside folder **gallery** (sub folder thumbs excluded of course). Maybe we could integrate this script inside the `build-markup`
- I'm using this [Easy Image](https://www.npmjs.com/package/easyimage) for thumb generation. Is nice. I like.
- all  thumbnails have 350x350 (but we can change this on the new `make-thumbs.js`)
- added "View Gallery" button
- all images display on separate tab when you click

Check it out and le me know what you think.

Going to drink some beers now

simonmcmanus commented:
:+1: 

Your example json looks sensible. 

I think it's a good idea to keep the make-thumbs command separate from the build markup script for now. 

For the time being I would add a script into the package.json file, something like 

npm run photos

I've pulled it down, on the gallery page I still see some weird white space issues: 

![screen shot 2016-02-14 at 08 21 11](https://cloud.githubusercontent.com/assets/55853/13032680/165a098c-d2f4-11e5-931b-9cffe2ba46f5.png)

A couple of other v minor tweaks:

The View gallery button does not need to be a button or CTA, it can just be a plain old left aligned link.

If we are going to have three images on the homepage the three columns should line up with the three columns for the talks on the homepage.  Just some margins need tweaking I think. 

Should be good to merge after that.

rdquintas committed:
adding script npm run photos

rdquintas committed:
fixing css white space and other minor adjustments

rdquintas committed:
css fixes

rdquintas committed:
updating README

rdquintas committed:
css media-query minor bug

iancrowther commented:
Is nice, merge conflicts tho

rdquintas commented:
Hi @simonmcmanus 
Here's another update
- fixed weird white space issues
- changed "view gallery" button to link
- on the homepage the three columns should line up with the three columns for the talks on the homepage (mind you that they are aligned to the outtermost left and right borders, but the pics are then spaced automatically - so it is not pixel-perfect). Let me know what you think
- add a script into the package.json file, something like npm run photos

@iancrowther : where/how can I help with the merge conflicts ?

simonmcmanus commented:
I suspect the merge conflicts are because the map pictures are being fetched each time and git thinks they have changed (despite the fact that they look the same)

This is not to do with your change @rdquintas - it was something that needed fixing anyway.

To get round this I think we should remove the map generation from the 'npm run build' command so that we only generate a new map when the map actually changes. Probably by creating npm run map command

I think we may still find the same problem with the appCache.manifest as it's generated and timestamped every time we build the html. I think it's worth moving the maps out of npm build first, resetting your map pictures and seeing if that resolved the merge conflicts. 

I can give this a go later but thought it worth writing down in case someone wants to have a go :)

From a quick v look at the diff it looks like your still committing the big pictures with the small ones and your import script just looks at that folder to generate the thumbnails. 

In order to get this PR merged I would just include the thumbs in this PR, we can work on the import script in a separate branch/PR. 

I will try to pull down the code and have a play later today. 

Cheers @rdquintas!

rdquintas commented:
Hi @simonmcmanus 
Ok no problem, I will just include the thumbnails and remove the originals from the repo.

But still have a question:
what about when you click on the thumbs ? I thought you wanted the original images to  open on a separate tab. If I remove the original images, then we won't have this.
Should I remove the `href` from the thumbs as well then ?

simonmcmanus commented:
All the pics are being pulled in from the web right?

Can we just link to their original url?

Cheers

On 15 February 2016 at 08:54, Ricardo Quintas notifications@github.com
wrote:

> Hi @simonmcmanus https://github.com/simonmcmanus
> Ok no problem, I will just include the thumbnails and remove the originals
> from the repo.
> 
> But still have a question:
> what about when you click on the thumbs ? I thought you wanted the
> original images to open on a separate tab. If I remove the original images,
> then we won't have this.
> Should I remove the href from the thumbs as well then ?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/lnug/lnug.github.io/pull/108#issuecomment-184119037.

## 

Simon McManus

DotJS Ltd - Node Consultancy

rdquintas commented:
Yes they are.... manually copied by me into the folder
:blush: 

But I will change that. No worries.
Sorry for getting confused.

Will use instead the JSON file as suggested, with all the info for each pic (URL, author, etc....) and then use it to build the markup.

ok ?

Cheers @simonmcmanus

rdquintas committed:
fixing

rdquintas committed:
merging maps-breakout

iancrowther commented:
@simonmcmanus is this ready to merge? I'm getting a bit lost in this stream of consciousness :-)

simonmcmanus commented on bin/archive.html:
```diff
@@ -0,0 +1 @@
+undefined
```
this file isn't needed.

simonmcmanus commented on bin/archive.json:
```diff
@@ -0,0 +1,8 @@
+{
```
this file isn't needed.

simonmcmanus commented on bin/code-of-conduct.html:
```diff
@@ -0,0 +1 @@
+undefined
```
this file isn't needed.

simonmcmanus commented on bin/code-of-conduct.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "page": "code-of-conduct"
+}
```
this file isn't needed.

simonmcmanus commented on bin/contact.html:
```diff
@@ -0,0 +1 @@
+undefined
```
this file isn't needed. the bin folder should only contain javascript files. I think you may have run a build in the bin directory and then committed all the files it generated.

simonmcmanus commented on bin/contact.json:
```diff
@@ -0,0 +1 @@
+{}
```
no needed.

simonmcmanus commented on code-of-conduct.html:
```diff
@@ -96,6 +96,7 @@
       <object class="footer-logo" data="images/lnug-logo-monochrome.svg" type="image/svg+xml"></object>
       <ul class="footer-links">
         <li><a href="./related-meetups.html" title="related meetups">Related Meetups</a></li>
+        <li><a href="./image-gallery.html" title="archive">Gallery</a></li>
```
title needs updating in layout

simonmcmanus commented on components/image-gallery (1)/image-gallery.html:
```diff
@@ -0,0 +1,5 @@
+<div class="single-image">
```
image-gallery (1)? Is this needed?

simonmcmanus commented on data/image-gallery.json:
```diff
@@ -0,0 +1,125 @@
+        "date": "20150101"
+    }]
+
+}
```
whats this file all about?

simonmcmanus commented:
@rdquintas  - Hope all the above makes sense, I think the functionality is not to far off but we need to remove the files which are not doing anything as they will cause confusion. 

Happy to jump on the branch and remove them if it will help.

rdquintas commented:
Hi @simonmcmanus yep they all make sense.
Sorry for not having replied earlier.
Been busy with go-live stuff here on my other project.

Will fix this today.

rdquintas committed:
deleting unnecessary files from folder bin

rdquintas committed:
fixing TITLE and removing image-gallery(01) folder

rdquintas commented:
Hi @simonmcmanus should be allright now.
Sorry for the mess with the `bin` folder. My bad.

simonmcmanus commented on css.css:
```diff
@@ -3,437 +3,516 @@
  *
```
All the white space changes to this css file make it really hard to see what (if anything) you have changed. 

If you think the whitespace needs correcting that should be done as a separate PR.

simonmcmanus commented on image-gallery.html:
```diff
@@ -0,0 +1,196 @@
+  </header>
+
+  <div id="container"><section id="gallery"><div class="single-image">
+    <a target="_blank" href="/images/gallery/11049186_1572780413006207_1571944090_n.jpg">
```
can this just point to the origin of the image? So if pulled from flickr or instagram just link to the image on instagram?

simonmcmanus commented on image-gallery.html:
```diff
@@ -0,0 +1,196 @@
+
+  <div id="container"><section id="gallery"><div class="single-image">
+    <a target="_blank" href="/images/gallery/11049186_1572780413006207_1571944090_n.jpg">
+        <img class="lazy" src="/images/gallery/thumbs/11049186_1572780413006207_1571944090_n.jpg">
```
do we still need a class of lazy on here?

simonmcmanus commented on image-gallery.html:
```diff
@@ -0,0 +1,196 @@
+</div>
+<div class="single-image">
+    <a target="_blank" href="/images/gallery/11084735_1427020997593875_864466577_n.jpg">
+        <img class="lazy" src="/images/gallery/thumbs/11084735_1427020997593875_864466577_n.jpg">
```
how are you naming these files? Just wondering how this will work when you start automatically pulling new images into the folder.

rdquintas committed:
fixing CSS, un-doing indent and removing white spaces

rdquintas committed:
removing references to lazy class

rdquintas committed:
building markup

rdquintas committed:
removing references to lazy class

rdquintas committed:
using gallery.json as the source to build the gallery markup. all images, except thumbs, are now externally stored

rdquintas commented:
Hi @simonmcmanus 
Hope all is ok now.
Please check and let me know.

Markup for the gallery is now being built using `gallery.json` as suggested.
For now, this file was manually updated, but in the near future I want to improve it and automate its process.
But like you said, should be handled on separate PR.

I have also fixed the CSS, that is, un-doed all that "noise" caused by unnecessary indent, white space, etc...

Cheers

rdquintas committed:
small fix the random javascript logic

simonmcmanus commented on README.md:
```diff
@@ -45,7 +45,24 @@ Once you have built the site confirm it works locally before pushing to master a
 
 Check the package.json file for a full list of commands you can run.
 
+For the image-gallery, just dump all your photos inside the folder **images/gallery**.
```
are these instructions still correct? dont you just update the json file now?

rdquintas replied:
You're right, they aren't 100% accurate
Will fix this.

simonmcmanus commented on bin/photos.js:
```diff
@@ -0,0 +1,31 @@
+var easyimage = require('easyimage')
+var fs = require('fs')
+
+fs.readdir('./images/gallery', function (err, items) {
```
looks like this is still going over the images gallery folder. 
Can this just look at the json file for any url without a thumbnail and update the json?

That way we can pull the urls into the json file and automate the whole process.

rdquintas replied:
@simonmcmanus 
Yes, that's correct.
The `photo.js` is not yet automated, hence it is still pointing to images/gallery to build the thumbnails. No integration with json yet.

The reason is simple -  I haven't had the time to implement this yet. :blush: 
I was planning to do this on a separate PR in the next couple days.

Basically, we currently have a gallery that works but is not automated.

simonmcmanus replied:
So the current flow is drop the big images into the gallery folder, run the photos script to generate the thumbnails and then remove the large images again? 

It seems simpler to just add the image urls into the json file. 

I can imagine a situation where someone forgets to delete the original images and we end up with lots of large images in the repo. 

If this will be fixed in a subsequent PR I don't mind merging but only as a temporary measure.

rdquintas replied:
**_So the current flow is drop the big images into the gallery folder, run the photos script to generate the thumbnails and then remove the large images again?_**
Correct.

**_It seems simpler to just add the image urls into the json file._**
Agree.
But the problem is in generating thumbs this way. Guess we would need to use the API for each platform where we're pulling the images from (one for twitter, another for Instagram, and so on...)
I also guess the module for generating thumbs doesn't work with URLs. 
You still need to provide it with a folder where to get the images from. 
But maybe I'm wrong, as I haven't fully analysed this.
One thing is for sure, the solution to automate this will require more thought and effort. 
But I can only dive into that on the next couple of days.
Maybe we can discuss this next Wednesday at the event, and you could give me your input on this.

**_I can imagine a situation where someone forgets to delete the original images and we end up with lots of large images in the repo._**
Agree. 
But thought we could go with a quick a simple gallery for now, and automate it latter.
Just guessing the risk you're mentioning wouldn't be that big after all. 

Cheers

rdquintas committed:
fixing README file

rdquintas commented:
Hi @simonmcmanus I have fixed the README

simonmcmanus commented:
Hi @rdquintas - sorry for the delay here. I will try to look this morning.
